revocation/ocsp/ocsp.go

```diff
@@ -0,0 +1,293 @@
+	}
+
+	var resp *http.Response
+	if base64.URLEncoding.EncodedLen(len(ocspRequest)) >= 255 {
+		reader := bytes.NewReader(ocspRequest)
+		resp, err = opts.HTTPClient.Post(server, "application/ocsp-request", reader)
+	} else {
+		encodedReq := base64.URLEncoding.EncodeToString(ocspRequest)
+		var reqURL string
+		reqURL, err = url.JoinPath(server, encodedReq)
+		if err != nil {
+			return nil, OCSPCheckError{Err: err}
+		}
+		resp, err = opts.HTTPClient.Get(reqURL)
+	}
+
+	if err != nil {
+		var urlErr *url.Error
+		if errors.As(err, &urlErr) && urlErr.Timeout() {
+			return nil, TimeoutError{}
+		}
+		return nil, OCSPCheckError{Err: err}
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
+		return nil, fmt.Errorf("failed to retrieve OSCP: response had status code %d", resp.StatusCode)
+	}
+
+	body, err := io.ReadAll(io.LimitReader(resp.Body, ocspMaxResponseSize))
+	if err != nil {
+		return nil, OCSPCheckError{Err: err}
+	}
+
+	switch {
+	case bytes.Equal(body, ocsp.UnauthorizedErrorResponse):
+		return nil, OCSPCheckError{Err: errors.New("OSCP unauthorized")}
+	case bytes.Equal(body, ocsp.MalformedRequestErrorResponse):
+		return nil, OCSPCheckError{Err: errors.New("OSCP malformed")}
+	case bytes.Equal(body, ocsp.InternalErrorErrorResponse):
+		return nil, OCSPCheckError{Err: errors.New("OSCP internal error")}
+	case bytes.Equal(body, ocsp.TryLaterErrorResponse):
+		return nil, OCSPCheckError{Err: errors.New("OSCP try later")}
+	case bytes.Equal(body, ocsp.SigRequredErrorResponse):
+		return nil, OCSPCheckError{Err: errors.New("OSCP signature required")}
+	}
+
+	return ocsp.ParseResponseForCert(body, cert, issuer)
+}
+
+func errToServerResult(err error) *base.ServerResult {
+	if err == nil {
+		return &base.ServerResult{
+			Result: base.OK,
+			Error:  nil,
+		}
+	} else if errors.Is(err, NoOCSPServerError{}) || errors.Is(err, PKIXNoCheckError{}) {
+		return &base.ServerResult{
+			Result: base.NonRevokable,
+			Error:  err,
+		}
+	} else if errors.Is(err, RevokedError{}) {
+		return &base.ServerResult{
+			Result: base.Revoked,
+			Error:  err,
+		}
+	}
+	// Includes OCSPCheckError, UnknownStatusError, base.InvalidChainError, and
+	// TimeoutError
+	return &base.ServerResult{
+		Result: base.Unknown,
+		Error:  err,
+	}
+}
+
+func serverResultsToCertRevocationResult(serverResults []*base.ServerResult) *base.CertRevocationResult {
+	if len(serverResults) == 1 {
+		return &base.CertRevocationResult{
+			Result:        serverResults[0].Result,
+			ServerResults: []*base.ServerResult{serverResults[0]},
+		}
+	}
+	currResult := base.OK
+	for _, serverResult := range serverResults {
+		switch serverResult.Result {
+		case base.Revoked:
+			currResult = base.Revoked
+		case base.Unknown:
+			if currResult != base.Revoked {
+				currResult = base.Unknown
+			}
+		case base.NonRevokable:
+			if currResult != base.Revoked && currResult != base.Unknown {
+				currResult = base.NonRevokable
+			}
+		}
```

Comment:
This logic is based upon objects that we have created, so there won't be an RFC associated with it. However, the logic here is intended to determine the correct result for a cert by looking to see the most significant Result from the server results (i.e. by level of importance, with a Revoked result being the most significant, Unknown being secondary only to a Revoked result, NonRevokable indicating no possible response, and lastly OK indicating no errors). It may be worth documenting this in the specification, or potentially just noting it in the definition for the Result of a CertRevocationResult

Reply:
> It may be worth documenting this in the specification, or potentially just noting it in the definition for the Result of a CertRevocationResult

I see. Yeah, I agree, I think we should document this logic design in both our spec and in the code.

Reply:
@kody-kimberl Cool. Could you open an issue at https://github.com/notaryproject/notaryproject/issues for this? just in case we forget to update the spec later.

Reply:
Update: After further discussions, we have decided to remove this waterfall logic. In the case of multiple results, the last result will be used since all of the previous results should be unknown (considering an error should have occurred). Thus, no change is needed to the spec